src/tqec/compile/specs/library/_utils.py

```diff
@@ -1,107 +1,262 @@
+        )
     )
-    final_plaquettes = Plaquettes(
+    if repetitions is not None:
+        plaquettes = plaquettes.repeat(repetitions)
+    return plaquettes
+
+
+def _build_regular_block(
+    builder: PlaquetteBuilder,
+    temporal_basis: Literal["X", "Z"],
+    x_boundary_orientation: Literal["VERTICAL", "HORIZONTAL"],
+    repetitions: LinearFunction = _DEFAULT_BLOCK_REPETITIONS,
+) -> CompiledBlock:
+    layers = [
+        _build_plaquettes_for_rotated_surface_code(
+            builder,
+            x_boundary_orientation,
+            temporal_basis=temporal_basis,
+            data_init=init,
+            data_meas=meas,
+            repetitions=repeat,
+        )
+        for init, meas, repeat in [
+            (True, False, None),
+            (False, False, repetitions),
+            (False, True, None),
+        ]
+    ]
+    return CompiledBlock(template=QubitTemplate(), layers=layers)
+
+
+def _build_substitution_in_time_direction(
+    pipe_spec: PipeSpec, plaquette_builder: PlaquetteBuilder
+) -> Substitution:
+    # substitute the final layer of the src block
+    src = {
+        -1: _build_plaquettes_for_rotated_surface_code(
+            plaquette_builder,
+            _get_x_boundary_orientation(pipe_spec.spec1),
+        )
+    }
+    # substitute the first layer of the dst block
+    dst = {
+        0: _build_plaquettes_for_rotated_surface_code(
+            plaquette_builder,
+            _get_x_boundary_orientation(pipe_spec.spec2),
+        )
+    }
+    return Substitution(src, dst)
+
+
+def _build_plaquettes_for_substitution(
+    builder: PlaquetteBuilder,
+    substitution_side: PlaquetteSide,
+    x_boundary_orientation: Literal["VERTICAL", "HORIZONTAL"],
+    *,
+    temporal_basis: Literal["X", "Z"] | None = None,
+    data_init: bool = False,
+    data_meas: bool = False,
+    repetitions: LinearFunction | None = None,
+) -> Plaquettes:
+    p1, p2 = _build_plaquette_for_different_basis(
+        builder,
+        x_boundary_orientation,
+        temporal_basis=temporal_basis,
+        data_init=data_init,
+        data_meas=data_meas,
+        init_meas_only_on_side=substitution_side,
+    )
+    mapping: dict[int, Plaquette]
+    if substitution_side == PlaquetteSide.LEFT:
+        mapping = {
+            1: p1.project_on_boundary(PlaquetteOrientation.UP),
+            7: p2,
+            8: p1,
+        }
+    elif substitution_side == PlaquetteSide.UP:
+        mapping = {
+            2: p2.project_on_boundary(PlaquetteOrientation.RIGHT),
+            5: p2,
+            6: p1,
+        }
+    elif substitution_side == PlaquetteSide.RIGHT:
+        mapping = {
+            4: p1.project_on_boundary(PlaquetteOrientation.DOWN),
+            11: p1,
+            12: p2,
+        }
+    else:  # substitution_side == PlaquetteSide.DOWN
+        mapping = {
+            3: p2.project_on_boundary(PlaquetteOrientation.LEFT),
+            13: p1,
+            14: p2,
+        }
```

Comment:
These two branches will break if any `Qubit4WayJunctionTemplate` is used. In general, there is a long-standing problem here which is "how to get the correct plaquette indices when performing substitution". We can leave the code like that, just adding a note or a TODO to be sure that we address that issue eventually.

Reply:
Yes, in my plan, more substitution rules can be added as branched to `_build_substitution_in_space` and this one(`_build_substitution_in_space_with_regular_cubes`) should only work for regular cube connections.

Reply:
I would rather advocate for an automated way of performing the substitution, but I am not sure it is doable in general (and I am sure it will require more code for this specific case), so I am fine with the current implementation.

Reply:
Yes, we may find out some algorithm to automate the substitution one day. But now I'm not aware of a simple way to do that.